handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -1447,65 +1484,224 @@ private static ByteBuffer toByteBuffer(ByteBuf out, int index, int len) {
+                            // ensure we flush any queued data..
+                            wrap(ctx, inUnwrap);
+                        } catch (Throwable e) {
+                            taskError(e);
+                            return;
                         }
-                    } catch (Exception e) {
-                        ctx.fireExceptionCaught(e);
-                    } finally {
-                        latch.countDown();
-                    }
+
+                        // Flush now as we may have written some data as part of the wrap call.
+                        forceFlush(ctx);
+
+                        // deliberate fall-through
+
+                    // We need more data so lets try to call decode again which will feed us with buffered data.
+                    case NEED_UNWRAP:
+                        tryDecodeAgain();
+                        break;
+
+                    // To make progress we need to call SSLEngine.wrap(...) which may produce more output data
+                    // that will be written to the Channel.
+                    case NEED_WRAP:
+                        try {
+                            wrapNonAppData(ctx, inUnwrap);
+                        } catch (Throwable e) {
+                            taskError(e);
+                            return;
+                        }
+                        // Flush now as we may have written some data as part of the wrap call.
+                        forceFlush(ctx);
+
+                        // Now try to feed in more data that we have buffered.
+                        tryDecodeAgain();
+                        break;
+                    default:
+                        // Should never reach here as we handle all cases.
+                        throw new AssertionError();
                 }
-            });
+            } catch (Throwable cause) {
+                handleException(cause);
+            }
+        }
 
-            boolean interrupted = false;
-            while (latch.getCount() != 0) {
-                try {
-                    latch.await();
-                } catch (InterruptedException e) {
-                    // Interrupt later.
-                    interrupted = true;
+        @Override
+        public void run() {
+            try {
+                task.run();
+                if (engine.getHandshakeStatus() == HandshakeStatus.NEED_TASK) {
+                    final Runnable task = engine.getDelegatedTask();
+                    // We only call the method when NEED_TASK is returned so we must have a task!
+                    assert task != null;
+                    delegatedTaskExecutor.execute(new SslTask(task, inUnwrap));
+                } else {
+                    if (ctx.executor().inEventLoop()) {
+                        resumeOnEventExecutor();
+                    } else {
+                        ctx.executor().execute(new Runnable() {
```

Comment:
Could make this `Runnable` a constant to avoid allocation?

It would be kind of nice for there to be an `EventExecutor.runInEventLoop(Runnable)` for cases like this :)

Reply:
@njhill as this is only done as part of the handshake and also it depends on `inUnwrap` I see no point of doing so.